SalesforceCommerceCloud/pwa-kit
# Make `commerce-sdk-react` Work Server-Side 💪🏻

# Description

This the server-side support for `react-query` being merged into the SDK in this #724. It was important that we prove that the `commerce-sdk-react` hooks work was compatible. In this PR I make the required changes to ensure that any hooks used via the `commerce-sdk-react` module are executed server side. 

I have left comments on part of this PR that I would like feedback on. So please pay special attention to those. 

Why didn't this work after imply updating the sdk? 

The reason this didn't work, is the `commerce-sdk-react` provider was itself using a react query provider. This provider was in conflict with the one supplied by the SDK. This meant that any hooks that we used used the client in the first provider it could find in its component hierarchy. This was the one defined in the `commerce-sdk-react`. Unfortunately the SDK does not know anything about this one and isn't able to track and execute the queries made by the hooks within it. This is the reason that it was required to be removed.

# Changes

- Remove the interfering ReactQueryProvider and client that was being defined in the SCAPI hooks provider.
- Updated the test application to use the new `withReactQuery` HoC from the SDK
- Added a section to the `commerce-sdk-react` readme on how its used in and out of a PWA-Kit application.
- Updated SDK to allow the `withReactQuery` to consume options for setting the clients configuration.
- Added default client configuration to never retry on the server. <-- Need feedback on this.

# How to Test-Drive This PR

- Checkout this branch.
- Run `npm ci`
- Run `npm start`
- Your browser will open, once it's loaded goto this url --> http://localhost:3000/products?mobify_server_only=1
- You shouldn't see a loading screen as you previously did, but instead you'll see the JSON data for the products requested immediately.

## General


## TODOs

- [x] Fix failing tests in `commerce-sdk-react`
